service/src/java/org/apache/hive/service/auth/HttpAuthService.java

```diff
@@ -0,0 +1,265 @@
+  private final CookieSigner signer;
+  private final String cookieDomain;
+  private final String cookiePath;
+  private final int cookieMaxAge;
+  private final boolean isCookieSecure;
+  private final String authCookieName;
+  
+  public HttpAuthService(String cookieDomain, String cookiePath, int cookieMaxAge, boolean isCookieSecure, 
+      String authCookieName) {
+    String secret = Long.toString(RAN.nextLong());
+    this.signer = new CookieSigner(secret.getBytes());
+    this.cookieMaxAge = cookieMaxAge;
+    this.cookieDomain = cookieDomain;
+    this.cookiePath = cookiePath;
+    this.authCookieName = authCookieName;
+    this.isCookieSecure = isCookieSecure;
+  }
+
+  public Cookie signAndCreateCookie(String cookieToken) {
+    return createCookie(signer.signCookie(cookieToken));
+  }
+
+  /**
+   * Generate a server side cookie given the cookie value as the input.
+   * @param str Signed input string token.
+   * @return The generated cookie.
+   */
+  public Cookie createCookie(String str) {
+    Cookie cookie = new Cookie(authCookieName, str);
+    cookie.setMaxAge(cookieMaxAge);
+    if (cookieDomain != null) {
+      cookie.setDomain(cookieDomain);
+    }
+    if (cookiePath != null) {
+      cookie.setPath(cookiePath);
+    }
+    cookie.setSecure(isCookieSecure);
+    return cookie;
+  }
+
+  /**
+   * Validate the request cookie. This function iterates over the request cookie headers
+   * and finds a cookie that represents a valid client/server session. If it finds one, it
+   * returns the client name associated with the session. Else, it returns null.
+   * @param request The HTTP Servlet Request send by the client
+   * @return Client Username if the request has valid HS2 cookie, else returns null
+   * @throws UnsupportedEncodingException in case of wrong encoding
+   */
+  public String validateCookie(HttpServletRequest request) throws UnsupportedEncodingException {
+    // Find all the valid cookies associated with the request.
+    Cookie[] cookies = request.getCookies();
+    if (cookies == null) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("No valid cookies associated with the request {}", request);
+      }
+      return null;
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Received cookies: {}", toCookieStr(cookies));
+    }
+    return getClientNameFromCookie(cookies);
+  }
+
+  /**
+   * Retrieves the client name from cookieString. If the cookie does not
+   * correspond to a valid client, the function returns null.
+   * @param cookies HTTP Request cookies.
+   * @return Client Username if cookieString has a HS2 Generated cookie that is currently valid.
+   * Else, returns null.
+   */
+  private String getClientNameFromCookie(Cookie[] cookies) {
+    // Current Cookie Name, Current Cookie Value
+    String currName;
+    String currValue;
+
+    // Following is the main loop which iterates through all the cookies send by the client.
```

Comment:
nit: send -> sent
I know it's not a new code in this PR but it's easy to fix

Reply:
Done